### PR TITLE
spanconfigsqlwatcher: add a `system.protected_ts_records` decoder

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds a decoder to decode rows from the
`system.protected_ts_records` table. This will be used by the
rangefeed we setup in the SQLWatcher, to decode rangefeed updates
and subsequently feed them into the SQLTranslator. The `target` column
of the system table is the only column of interest since it will
determine whether we ask the translator to generate `SystemSpanConfig`s
or regular `SpanConfigs`. More information about this will be added
in follow up PRs that use this decoder.

This change also instructs gogoproto to generate an equality method
for `roachpb.TenantID`, to allow for equality checks in the decoder
test.

Informs: #73727

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
@dt adding you for the small change to `roachpb.TenantID`

**Comment by dt** [comment]
> for the small change to roachpb.TenantID
I'll defer to @tbg or @RaduBerinde if we want the proto message to be comparable (I guess the go struct already is so I don't see why not, but we treat TenantID with care so let's check)

**Review by RaduBerinde — COMMENTED** [review]
Same here, I don't see why not.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74913#-:-MtezxjO9Bh3MXxzmSLe:b-z1sh34)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @dt)

**Comment by tbg** [comment]
That seems ok.

File: pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder.go:54]: Throw in an errors.Assertion here for the length of datums decoded.

File: pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go:51]: Instead of writing to the real table in this unit test, it'd be better isolated if using a scratch table instead: 
https://github.com/cockroachdb/cockroach/blob/89f8aba84346fcae27c55dc253a3f574c9eb8e2e/pkg/spanconfig/spanconfigkvsubscriber/span_config_decoder_test.go#L40-L42

Given this test should only care about the decoding routines, I would suggest writing to the scratch table in this way directly, instead of depending on jobs or the protected ts subsystem to write to the real table (and also needing to release the record). Also, since the column is nullable, worth having a nil check (an empty proto should be returned -- something we'll need to explicitly handle).

Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go:51]: Hmm doesn't it make sense to test this with the `jobsprotectedts.MakeRecord`, `Protect`, and `Release` methods that are actually going to be used to interact with this table by all PTS consumers? We use this pattern in many other PTS tests as well. In theory, `target` is a free-form bytes column and the decoder could be doing something different from what the actual system writes to the column, so it seems a little scary to decouple it from how a pts record will actually be written?

File: pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder.go:54]: done!

File: pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go
Comment (irfansharif) [reply pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go:51]: Hm, I'm not sure, but I see your point. I was coming at it to minimize the test/make it fast enough to also then suggest writing a benchmark for it ("TestProtectedTimestampDecoder only cares about decoding"). Minimizing the test also prevents it from failing when another subsystem is buggy (`Protect`/`Release`); the nice thing about talking through this table _is_ that these sub-systems are decoupled, I think it's fine/sane to have them decoupled in test code as well. I'll defer to you and not hold up the PR.

**Review by irfansharif — APPROVED** [review]

File: pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go
Comment (adityamaru) [reply pkg/spanconfig/spanconfigsqlwatcher/protectedtsdecoder_test.go:51]: > Minimizing the test also prevents it from failing when another subsystem is buggy

Coming from bulkio land where we write almost all tests that rely on the jobs subsystem, this is something I've just taken for granted before heh. The benchmarking is a good point, I'm leaning towards merging this for now since I have CI happy, and cleaning it up during stability if we decide to benchmark some of these pieces.

**Comment by adityamaru** [comment]
TFTR!

bors r=irfansharif

**Pull request merged**
